Additional connector view setting

EdoFro opened the pull request:
Hi Dimitry,
I had some spare time today and solved a feature request that I made some time ago 
( **Additional Connector view setting (combination of 'for selection' + 'only arrows')** #741)


It's a new connector view setting:
![image](https://github.com/freeplane/freeplane/assets/35700575/59c2f3ef-3946-4f55-9b7c-1a0030526e29)

it combines "only arrows" and "for selection":  
### it shows the connectors of the selected node(s) and only the arrow heads for the other nodes.

Example:
[Test new Connector View.zip](https://github.com/freeplane/freeplane/files/13462624/Test.new.Connector.View.zip)


![image](https://github.com/freeplane/freeplane/assets/35700575/3114d4c8-df01-4bb9-b548-0ae095ca634f)

To compare: the same map with view setting "always":
![image](https://github.com/freeplane/freeplane/assets/35700575/d2506be9-8cd0-4955-a9f1-d4d64b71b586)

### I also changed the arrows to be different when entering or leaving a node:

this:
![image](https://github.com/freeplane/freeplane/assets/35700575/fa5456da-73e3-43ce-bacf-0655eb914aea)

becomes this:
![image](https://github.com/freeplane/freeplane/assets/35700575/3c910101-3e6e-4959-a201-88b96d316d51)


Hope it helps,

edo

EdoFro committed:
adds show_connectors.for_selection_and_arrows

additional view setting

EdoFro committed:
different start and end arrows for connectors

EdoFro committed:
Merge branch 'freeplane:1.11.x' into Additional-Connector-view-setting

EdoFro commented:
Sample video:

https://github.com/freeplane/freeplane/assets/35700575/59d4a527-52ac-4d58-90bc-58fb30e0658e

euu2021 commented:
This is useful. Thanks!

EdoFro committed:
Changes form of "no arrow" end

changes teh connectors end that has no arrow to a diamond when connector's line is not shown.

I found that the previous form was not so easy to see when changing the mind map's view zoom factor.

EdoFro commented:
I changed the form of the connector's end that has no arrow to a diamond (when connector's line is not shown).

I found that the previous form was not so easy to see when changing the mind map's view zoom factor.

Now it looks so:
![image](https://github.com/freeplane/freeplane/assets/35700575/f9e4c221-4daa-4b7a-b80e-76e41db3f81a)

previously:
![image](https://github.com/freeplane/freeplane/assets/35700575/3c910101-3e6e-4959-a201-88b96d316d51)

EdoFro committed:
Merge branch 'freeplane:1.11.x' into Additional-Connector-view-setting

EdoFro committed:
Merge branch 'freeplane:1.11.x' into Additional-Connector-view-setting

biswajit-goswami commented:
@EdoFro  is it possible also to have an option to force the connectors to target the center of the node (source, destination) ?

dpolivaev commented:
I would rather call it "only arrows for selection"

EdoFro commented:
> I would rather call it "only arrows for selection"

Ok 👍🏼 
I don't have a better proposal. I think that when added to a preview version, the other users can ask if they don't undertand it and propose better wording.

EdoFro commented:
> @EdoFro is it possible also to have an option to force the connectors to target the center of the node (source, destination) ?

I don't know. I didn't investigate so far. Maybe for a next time.

EdoFro committed:
changes menu text to "only arrows for selection"

As suggested here:
https://github.com/freeplane/freeplane/pull/1548#issuecomment-1854506697

EdoFro committed:
makes arrow point a littlier "arrowier"

EdoFro commented:
I added 2 new commits:
1. changes the menu command to ["only arrows for selection"](https://github.com/freeplane/freeplane/pull/1548/commits/8affaa6053fcd4722e32742409afc46a5dccb30c)
2. [makes arrow head a littlier "arrowier"](https://github.com/freeplane/freeplane/pull/1548/commits/5f4c158801a78cfd9a17f0f223284423255dc04d)

this is how it looks now:
1. showing connector lines:
![image](https://github.com/freeplane/freeplane/assets/35700575/3ca84603-666f-4866-a3a4-5c96374ba1ed)

2. hiding connector lines and showing only arrow heads an connector tails:
![image](https://github.com/freeplane/freeplane/assets/35700575/e687ec58-2492-4a6b-bfea-8dbf88bb46d2)

3. "only arrows for selection"
![image](https://github.com/freeplane/freeplane/assets/35700575/a761773a-500e-4c0c-8c0b-2450db7c8bdc)

EdoFro committed:
Merge branch 'freeplane:1.11.x' into Additional-Connector-view-setting

EdoFro committed:
Merge branch 'freeplane:1.11.x' into Additional-Connector-view-setting

EdoFro commented:
I'm testing this PR into the new version.
There are no conflicts until now.

my first test with Code Explorer (and the new Connector View Setting):
![image](https://github.com/freeplane/freeplane/assets/35700575/73997208-e3f4-4b5e-a8cb-3c54709e4223)

Looks good!

dpolivaev commented:
Why do you like the so-called "diamonds" more than the arrows? I do not understand them at all.
I like the older version more where you used filled arrows for incoming connectors and non filled arrows for the outgoing connectors. Painting diamonds has not only added complexity to the code. It made the map not as clear to me as it was before the change.

EdoFro commented:
> Why do you like the so-called "diamonds" more than the arrows? I do not understand them at all. I like the older version more where you used filled arrows for incoming connectors and non filled arrows for the outgoing connectors. Painting diamonds has not only added complexity to the code. It made the map not as clear to me as it was before the change.

Ok. Here are three other options:

Normal "allways" view (as comparission base):
![image](https://github.com/freeplane/freeplane/assets/35700575/82fefe06-4e74-48bc-b9e9-f23941510541)

-----

## 1. Fill and direction
   - connector in:     filled arrow pointing in
   - connector out:  non filled arrow pointing out
   
![image](https://github.com/freeplane/freeplane/assets/35700575/4da0568e-6ba8-4af0-a6ef-54f305524676)

code in paintArrow method:
```java
    	final Polygon p = new Polygon();
    	if(isArrowsPoint) {
    		p.addPoint((to.x), (to.y));
    		p.addPoint((int) (to.x + dxn + arrowWidth * dyn), (int) (to.y + dyn - arrowWidth * dxn));
    		p.addPoint((int) (to.x + dxn - arrowWidth * dyn), (int) (to.y + dyn + arrowWidth * dxn));
    		p.addPoint((to.x), (to.y));
		g.fillPolygon(p);
    	} else {
		p.addPoint((int) (to.x + dxn), (int) (to.y + dyn));
    		p.addPoint((int) (to.x + arrowWidth * dyn), (int) (to.y - arrowWidth * dxn));
    		p.addPoint((int) (to.x - arrowWidth * dyn), (int) (to.y + arrowWidth * dxn));
		p.addPoint((int) (to.x + dxn), (int) (to.y + dyn));
    	}
    	g.drawPolygon(p);

```

-----

## 2.  same but "arrowier" : fill and direction
   - connector in: filled arrow pointing in
   - connector out: non filled arrow pointing out

![image](https://github.com/freeplane/freeplane/assets/35700575/251550bf-d9d4-4fa0-8d0f-d5b42846347b)

code in paintArrow method:
```java
    	final Polygon p = new Polygon();
    	if(isArrowsPoint) {
    		p.addPoint((to.x), (to.y));
    		p.addPoint((int) (to.x + dxn + arrowWidth * dyn), (int) (to.y + dyn - arrowWidth * dxn));
    		p.addPoint((int) (to.x + dxn * 0.8d), (int) (to.y + dyn * 0.8d));
    		p.addPoint((int) (to.x + dxn - arrowWidth * dyn), (int) (to.y + dyn + arrowWidth * dxn));
    		p.addPoint((to.x), (to.y));
		g.fillPolygon(p);
    	} else {
		p.addPoint((int) (to.x + dxn), (int) (to.y + dyn));
    		p.addPoint((int) (to.x + arrowWidth * dyn), (int) (to.y - arrowWidth * dxn));
		p.addPoint((int) (to.x + dxn * 0.2d), (int) (to.y + dyn * 0.2d));
    		p.addPoint((int) (to.x - arrowWidth * dyn), (int) (to.y + arrowWidth * dxn));
		p.addPoint((int) (to.x + dxn), (int) (to.y + dyn));
    	}
    	g.drawPolygon(p);

```

-----

## 3. only Fill
   - connector in: filled arrow pointing in
   - connector out: non filled arrow pointing **in**

![image](https://github.com/freeplane/freeplane/assets/35700575/809a59f5-a9af-4d6d-912f-01250c4b999b)


code in paintArrow method:
```java
    	final Polygon p = new Polygon();
	p.addPoint((to.x), (to.y));
	p.addPoint((int) (to.x + dxn + arrowWidth * dyn), (int) (to.y + dyn - arrowWidth * dxn));
	p.addPoint((int) (to.x + dxn - arrowWidth * dyn), (int) (to.y + dyn + arrowWidth * dxn));
	p.addPoint((to.x), (to.y));
    	if(isArrowsPoint)
		g.fillPolygon(p);
    	g.drawPolygon(p);

```

----



Tell me what you think and I commit the new code.

BR

edo

EdoFro commented:
Here is a 4th option:

- connector in: filled arrow pointing in
   (TBD: arrowier or simple triangle)
- connector out: short line

![image](https://github.com/freeplane/freeplane/assets/35700575/c38b9f2e-28d2-4493-96ae-18366bc3d975)


code in paintArrow method:
```java
    	final Polygon p = new Polygon();
    	if(isArrowsPoint) {
    		p.addPoint((to.x), (to.y));
    		p.addPoint((int) (to.x + dxn + arrowWidth * dyn), (int) (to.y + dyn - arrowWidth * dxn));
    		p.addPoint((int) (to.x + dxn * 0.8d), (int) (to.y + dyn * 0.8d));  // this line is needed if "arrowier" option is taken
    		p.addPoint((int) (to.x + dxn - arrowWidth * dyn), (int) (to.y + dyn + arrowWidth * dxn));
    		p.addPoint((to.x), (to.y));
		g.fillPolygon(p);
    	} else {
		p.addPoint((to.x), (to.y));
		p.addPoint((int) (to.x + dxn), (int) (to.y + dyn));
    	}
    	g.drawPolygon(p);
```

I think, I like this better.

It's your choice. I think all options are good ones.
Or maybe another idea?

BR

edo

EdoFro commented:
Idea:
If this options (and their codes) are ok for you, then we can create a poll in the discussions forum and use the opinions of other users to decide.
It can be a nice thing for the community.

In the meantime one of the options can be added to the preview version, but the stable version should have the winning option.

opinion?

dpolivaev commented:
Personally I like your first version more than any other:

<img width="249" alt="arrows" src="https://github.com/freeplane/freeplane/assets/1558052/c172298d-de2a-40a0-82b6-1d16e0102484">

Here I see filled triangles for incoming connectors and arrows consisting of 2 non-closed lines for outgoing connectors.

When I ask the community I ask people to deliver opinions and arguments, not votes. People who do decide what to do hear voices from the community but keep all freedoms to make decisions that make the makers happy. I usually do not have polls.

I would not try to implement all options, I think that they pictures are sufficient to have a discussion.

I want to make concrete suggestions about improvements in the code after we get the functionality right, but I am going to comment some things in the code right now so that you can see what I think.

dpolivaev commented on freeplane/src/main/java/org/freeplane/view/swing/map/link/AConnectorView.java:
```diff
@@ -55,6 +55,10 @@ protected MapView getMap() {
+		paintArrow(from, to, g, size, true);
+	}
+
+	protected void paintArrow(final Point from, final Point to, final Graphics2D g, final double size, final boolean isArrowsPoint) {
```
instead of making this method more complicated, I would rather implement 2 methods so that we can see what exactly they paint based on their names and that their complexity is not higher than the complexity of the original method.

I do not undestand the meaning of argument name`isArrowsPoint`, but instead of trying to find a better name for the argument I would rather eliminate the argument and use 2 methods calling the one you need.

Based on the purpose of the methods I would probably call them `paintIncomingConnectorArrow` and `paintOutgoungConnectorArrow`. These names do not say anything about the arrow form but they explain their purpose.

dpolivaev commented:
I implemented my ideas and sent you pull request https://github.com/EdoFro/freeplane/pull/3
Please review it. If you accept it, I will merge your PR to the main repo immediately.
Otherwise, we need to continue our discussions.

dpolivaev commented:
Preview 1.11.9_07

@EdoFro 
I have merged your branch with my changes.
Don't hesitate to comment and disagree if there is anything.

EdoFro commented:
Hi @dpolivaev,
Sorry I've been so absent and have taken so long to respond. The last few weeks have been very consuming for me. I have had some problems both at work and family level. Nothing serious, but with high uncertainty and has required my full attention.

I just installed the latest version. I will be reviewing it and if I have any comments to contribute I will let you know.

Best regards,

edo

euu2021 commented:
I don't know if I understood correctly, but the implemented behavior is different from the one shown in this example: https://github.com/freeplane/freeplane/pull/1548#issuecomment-1826168036

What I'm seeing here:
In a situation like this, with the option `View->View settings->Show connector lines->always`:
![image](https://github.com/freeplane/freeplane/assets/77707706/96696612-d968-489d-8c52-add06cbb646e)

When I choose the new option `View->View settings->Show connector lines->only arrows for selection,` I get:
![image](https://github.com/freeplane/freeplane/assets/77707706/48885955-b7b6-4454-8adf-a0999731d350)

But, looking at the Edo examples, I would expect to see the arrow head for the nodes that are not selected, and the full connector for the nodes that are selected. Like this:
![image](https://github.com/freeplane/freeplane/assets/77707706/8a7cfec7-c363-4f29-9210-39baaf16845a)

Did I misunderstand?

aaa1386 commented:
> But, looking at the Edo examples, I would expect to see the arrow head for the nodes that are not selected, and the full connector for the nodes that are selected. Like this:

This is a great suggestion. The map doesn't get cluttered with connecting lines. The selected node is the node I'm focused on. Showing the connecting lines gives me a better understanding of the node. I can easily adjust or delete it by right-clicking on the connecting line.
